### PR TITLE
Perturbation

### PR DESCRIPTION
**Pull request by pseegerer**
Option to compute analysis once in the beginning and not before every perturbation step.

### CODE REVIEW

**Comment by albermax** [comment]
Hi Philipp,

thank you for your effort! The algorithms looks correct now!

The biggest problem I see right now is, and sorry I did not notice it before, is that:

* it is not deterministic as the (random) perturbation are recalculated each time instead of extending the previous.
* one cannot expect to hold the whole dataset in memory.

Especially if one would like to do a larger analysis like on ImageNet this will be a problem.
I am not sure how one can change the code best and also don't want to interfere with your implementation ideas, but I guess the best solution is to change the looping. Instead of:
```python
for each steps:
  for each image:
    # do xyz
```

```python
for each image:
  for each step:
    # do xyz (reusing previous step's image)
```

There are few minors that should be fixed:

* It is important that we can compare the results to Samek et.al. to be confident that it works. Therefore in the example:
  * the whole training set should be used.
  * the APOC curves like in Samek et. al. should be calculated. It would be good to reproduce their plot, which I guess can be easily done by adding the score on the un-perturbed image in front of the scores. Then (scores-scores[0])/len(scores) = APOC. There is also a random analyzer to obtain the random ordering.
* The row with analysis plots in the notebook does never change. In fact this row seems redundant. It would be great, if next to the method title there would be an image of of the analysis and next to it a heatmap with the ranked regions to highlight the order to the user.
* Is there a possibility to invert the ordering of the flipping? Such an option would be good. It is used in http://iphome.hhi.de/samek/pdf/BinICANN16.pdf
* It would be good to have an option to specify a value range and in that case the perturbation gets clipped. Default it should not be applied.

Some questions:
* What happens if a user passes a larger step number than possible?
* What happens if regions do not fit an image?
* Is it possible to pass a model without softmax and are the scores then the pre-softmax activations?

Cheers,
Max

**Comment by pseegerer** [comment]
> Especially if one would like to do a larger analysis like on ImageNet this will be a problem.
> I am not sure how one can change the code best and also don't want to interfere with your implementation ideas, but I guess the best solution is to change the looping. Instead of:
> 
> ```python
> for each steps:
>   for each image:
>     # do xyz
> ```
> 
> ```python
> for each image:
>   for each step:
>     # do xyz (reusing previous step's image)
> ```
This will be done soon.

> There are few minors that should be fixed:
> 
> * It is important that we can compare the results to Samek et.al. to be confident that it works. Therefore in the example:
>   
>   * the whole training set should be used.

Done.
>   * the APOC curves like in Samek et. al. should be calculated. It would be good to reproduce their plot, which I guess can be easily done by adding the score on the un-perturbed image in front of the scores. Then (scores-scores[0])/len(scores) = APOC. There is also a random analyzer to obtain the random ordering.

Partially done. AOPC are calculated but not yet plotted in exactly the same way as in Samek et al (i.e. relative to random analyzer). Maybe we can discuss this later.
> * The row with analysis plots in the notebook does never change. In fact this row seems redundant. It would be great, if next to the method title there would be an image of of the analysis and next to it a heatmap with the ranked regions to highlight the order to the user.

Heatmap and ordering (as heatmap, too) are plotted next to perturbations.
> * Is there a possibility to invert the ordering of the flipping? Such an option would be good. It is used in http://iphome.hhi.de/samek/pdf/BinICANN16.pdf

It is possible by e.g.: `perturbation.aggregation_function = lambda x, axis: -np.mean(x, axis=axis)`. I added this example to the notebook (commented out though). What is possible too, is to make a deep copy of the original aggregation function and then invert it, like `perturbation.aggregation_function = lambda x, axis: -deepcopy(perturbation.aggregation_function)(x, axis=axis)`. Should I implement this or should I leave it as is for now?
> * It would be good to have an option to specify a value range and in that case the perturbation gets clipped. Default it should not be applied.

A variable `value_range = (minval, maxval)` can now be set in `Perturbation`.
> Some questions:
> 
> * What happens if a user passes a larger step number than possible?

All the regions are perturbated for the last iterations. Should I print a warning if this is the case?
> * What happens if regions do not fit an image?

It is padded to the appropriate size. Pad mode can be set by user, default is `reflect`.
> * Is it possible to pass a model without softmax and are the scores then the pre-softmax activations?

Yes, this should be possible.

File: innvestigate/tools/perturbate.py [hunk 1]
```diff
@@ -127,6 +139,10 @@ def perturbate_regions(self, x, perturbation_mask_regions):
             if region_mask:
                 x_perturbated[sample_idx, channel_idx, region_row, :, region_col, :] = self.perturbation_function(
                     region)
+
+                if self.value_range is not None:
+                    np.clip(x_perturbated, *self.value_range, x_perturbated)
+                    assert x_perturbated.min(), x_perturbated.max() == self.value_range
```
Comment (albermax) [inline innvestigate/tools/perturbate.py]: This does not need to be true!

File: examples/notebooks/mnist_perturbation.ipynb [hunk 2]
```diff
@@ -400,8 +338,12 @@
     "    \n",
     "    # Set up the perturbation analysis\n",
     "    perturbation = Perturbation(perturbation_function, region_shape=region_shape, in_place=False)\n",
+    "    \n",
+    "    # Comment out to invert the perturbation order\n",
+    "    # perturbation.aggregation_function = lambda x, axis: -np.mean(x, axis=axis)\n",
+    "    \n",
```
Comment (albermax) [inline examples/notebooks/mnist_perturbation.ipynb]: :-)

File: innvestigate/tools/perturbate.py (hunk 1 above)
Comment (pseegerer) [reply innvestigate/tools/perturbate.py]: You're right ;)

**Comment by albermax** [comment]
Great work! thanks!

Answer is between the lines:

> > Especially if one would like to do a larger analysis like on ImageNet this will be a problem.
> > I am not sure how one can change the code best and also don't want to interfere with your implementation ideas, but I guess the best solution is to change the looping. Instead of:
> > ```python
> > for each steps:
> >   for each image:
> >     # do xyz
> > ```
> > 
> > 
> > ```python
> > for each image:
> >   for each step:
> >     # do xyz (reusing previous step's image)
> > ```
> 
> This will be done soon.
> 
> > There are few minors that should be fixed:
> > 
> > * It is important that we can compare the results to Samek et.al. to be confident that it works. Therefore in the example:
> >   
> >   * the whole training set should be used.
> 
> Done.
> 
> > * the APOC curves like in Samek et. al. should be calculated. It would be good to reproduce their plot, which I guess can be easily done by adding the score on the un-perturbed image in front of the scores. Then (scores-scores[0])/len(scores) = APOC. There is also a random analyzer to obtain the random ordering.
> 
> Partially done. AOPC are calculated but not yet plotted in exactly the same way as in Samek et al (i.e. relative to random analyzer). Maybe we can discuss this later.
> 

That's fine. I can adjust this later.

> > * The row with analysis plots in the notebook does never change. In fact this row seems redundant. It would be great, if next to the method title there would be an image of of the analysis and next to it a heatmap with the ranked regions to highlight the order to the user.
> 
> Heatmap and ordering (as heatmap, too) are plotted next to perturbations.
> 
> > * Is there a possibility to invert the ordering of the flipping? Such an option would be good. It is used in http://iphome.hhi.de/samek/pdf/BinICANN16.pdf
> 
> It is possible by e.g.: `perturbation.aggregation_function = lambda x, axis: -np.mean(x, axis=axis)`. I added this example to the notebook (commented out though). What is possible too, is to make a deep copy of the original aggregation function and then invert it, like `perturbation.aggregation_function = lambda x, axis: -deepcopy(perturbation.aggregation_function)(x, axis=axis)`. Should I implement this or should I leave it as is for now?
> 

How about an argument that flips the ordering in the function compute_region_ordering?
To me that seems the best way.

> > * It would be good to have an option to specify a value range and in that case the perturbation gets clipped. Default it should not be applied.
> 
> A variable `value_range = (minval, maxval)` can now be set in `Perturbation`.
> 
> > Some questions:
> > 
> > * What happens if a user passes a larger step number than possible?
> 
> All the regions are perturbated for the last iterations. Should I print a warning if this is the case?
> 

I guess this is when the step number is smaller?
I think this is not the expected behaviour. I think it's better to not perturbate all the input in the last step but rather only the according region.

When it's larger it would be good to have a warning or an error to raise the awareness.

> > * What happens if regions do not fit an image?
> 
> It is padded to the appropriate size. Pad mode can be set by user, default is `reflect`.

Seems like a good solution to me!
> 
> > * Is it possible to pass a model without softmax and are the scores then the pre-softmax activations?
> 
> Yes, this should be possible.
